Not too urgent, but docs build is currently blocked until: https://github.com/agencyenterprise/neurotechdevkit/pull/124 merges

Error message:
```
Unexpected failing examples:
/home/circleci/project/docs/examples/plot_time_reverse.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/docs/examples/plot_time_reverse.py", line 62, in <module>
    true_scenario.render_layout()
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 894, in render_layout
    field = self.get_field_data("layer").astype(int)
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 339, in get_field_data
    return self.problem.medium.fields[field].data
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 110, in problem
    return self._problem
AttributeError: 'Scenario2_2D' object has no attribute '_problem'
```

--- a/docs/examples/plot_time_reverse.py
+++ b/docs/examples/plot_time_reverse.py
@@ -65,6 +65,7 @@ true_scenario.render_layout()
 # %%
 # ## Simulate the time-reverse scenario
 # Place a point source at the original target, and simulate a pulse.
+# The point source is visualized as a gray dot.
 
 # Reinitialize the scenario
 reversed_scenario = ndk.make(SCENARIO_NAME)

--- a/docs/examples/plot_time_reverse.py
+++ b/docs/examples/plot_time_reverse.py
@@ -1,0 +1,160 @@
+# -*- coding: utf-8 -*-
+"""
+Time-reverse simulation for phased array
+====================================================================
+
+The skull adds aberrations to the beam propagation; phased arrays can compensate
+for those by having different delays for each element, but estimating these
+delays can be challenging. One method shared by Sergio Jiménez-Gambín and
+Samuel Blackman is to run a "time reverse" simulation to calculate the
+necessary delays.
+
+This notebook shows an example of a "time reverse" simulation. The notebook
+sets up a scenario with a phased array source and a target and then runs
+a simulation with the source and target reversed to calculate the delays.
+Finally, it uses the calculated delays to perform a forward-time simulation.
+
+Note: In this notebook, we refer to the "true" target as the eventual brain
+region we would like to stimulate, and the "true" source as the placement of
+the ultrasound probes. We refer to the "reversed" or "simulated" target and
+point-source as the values defined in our simulation, which are reversed from
+the physical setup to help calculate values.
+"""
+
+# %%
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+import neurotechdevkit as ndk
+
+# Parameters
+SCENARIO_NAME = "scenario-2-2d-v0"
+NUM_ELEMENTS = 20
+ELEMENT_WIDTH = 1.2e-3
+
+
+# %%
+# Helper function to make the scenario with a PhasedArraySource
+def make_scenario(element_delays=None) -> ndk.scenarios.Scenario:
+    true_scenario = ndk.make(SCENARIO_NAME)
+
+    # define a phased-array source
+    default_source = true_scenario.get_default_source()
+    true_source = ndk.sources.PhasedArraySource2D(
+        element_delays=element_delays,
+        position=default_source.position,
+        direction=default_source.unit_direction,
+        num_elements=NUM_ELEMENTS,
+        pitch=default_source.aperture / NUM_ELEMENTS,
+        element_width=ELEMENT_WIDTH,
+        num_points=1000,
+    )
+
+    true_scenario.add_source(true_source)
+    return true_scenario
+
+
+# %%
+# ## Set up and visualize the forward scenario
+true_scenario = make_scenario()
+assert isinstance(true_scenario, ndk.scenarios.Scenario2D)
+true_scenario.render_layout()
+
+
+# %%
+# ## Simulate the time-reverse scenario
+# Place a point source at the original target, and simulate a pulse.
+
+# Reinitialize the scenario
+reversed_scenario = ndk.make(SCENARIO_NAME)
+# and reverse the source
+point_source = ndk.sources.PointSource2D(
+    position=true_scenario.target.center,
+)
+reversed_scenario.add_source(point_source)
+
+assert isinstance(reversed_scenario, ndk.scenarios.Scenario2D)
+reversed_scenario.render_layout()
+
+
+# %%
+result = reversed_scenario.simulate_pulse()
+assert isinstance(result, ndk.results.PulsedResult2D)
+result.render_pulsed_simulation_animation()
+
+
+# %% Calculate the time-reverse delays
+# We calculate how long it took for the point-source pulse to reach each of
+# the true array elements. Here, we coarsely approximate these delays by
+# finding the pressure argmax at each element's nearest-neighbor coordinates.
+
+# Map array elements onto the nearest pixels in our simulation
+def map_coordinates_to_indices(coordinates, origin, dx):
+    indices = np.round((coordinates - origin) / dx).astype(int)
+    return indices
+
+
+# Get the pressure time-series of these elements
+[true_source] = true_scenario.sources
+assert isinstance(true_source, ndk.sources.PhasedArraySource2D)
+element_indices = map_coordinates_to_indices(
+    true_source.element_positions,
+    reversed_scenario.origin,
+    reversed_scenario.dx,
+)
+pressure_at_elements = result.wavefield[element_indices[:, 0], element_indices[:, 1]]
+
+# Calculate the time of arrival for each element
+element_reverse_delays = np.argmax(pressure_at_elements, axis=1) * result.effective_dt
+plt.plot(element_reverse_delays, marker="o")
+plt.xlabel("element index")
+plt.ylabel("delay [s]")
+
+
+# %%
+# Visually inspecting the earlier scenario layout, these results seem reasonable.
+# The expected delay $$t_d$$ is approximately:
+#
+# $$t_d \approx \frac{||x_{source} - x_{target}||_2}{c_{water}} \approx \
+# \frac{0.07 \text{ m}}{1500 \text{ m/s}} \approx 47 \mu s$$
+#
+
+
+# %%
+# ## Use delays in forward-time simulation
+# Next, let's validate these delays by using them in a normal forward-time
+# simulation.
+# We simulate the original scenario, setting the pulse delays as calculated.
+
+# Elements that took longer to reach should now be pulsed first,
+# so we invert the values
+element_delays = element_reverse_delays.max() - element_reverse_delays
+
+true_scenario = make_scenario(element_delays=element_delays)
+result = true_scenario.simulate_pulse()
+assert isinstance(result, ndk.results.PulsedResult2D)
+result.render_pulsed_simulation_animation()
+
+
+# %%
+# The pulse should focus on the true target.
+
+
+# %%
+# ### Simulate steady-state
+# Another way to visualize the simulation is to check that the steady-state
+# pressure (within the skull) peaks near the target.
+
+# Re-initialize scenario to clear previous simulation
+true_scenario = make_scenario(element_delays=element_delays)
+steady_state_result = true_scenario.simulate_steady_state()
+assert isinstance(steady_state_result, ndk.results.SteadyStateResult2D)
+steady_state_result.render_steady_state_amplitudes()
+
+
+# %%
+# ## Future directions for improvement
+# The time-reverse simulation is not an exact solution for the forward-time
+# design. Other factors, like the angle of incidence at the boundary of two
+# materials, will be different in the time reverse vs forward-time.

--- a/src/neurotechdevkit/rendering/__init__.py
+++ b/src/neurotechdevkit/rendering/__init__.py
@@ -6,7 +6,7 @@ from ._animations import (
     save_animation,
     video_only_output,
 )
-from ._source import source_should_be_flat
+from ._source import source_should_be_flat, source_should_be_point
 from .layers import (
     SourceDrawingParams,
     draw_material_outlines,
@@ -42,6 +42,7 @@ __all__ = [
     "video_only_output",
     "display_video_file",
     "source_should_be_flat",
+    "source_should_be_point",
     "configure_matplotlib_for_embedded_animation",
     "save_animation",
 ]

--- a/src/neurotechdevkit/rendering/_source.py
+++ b/src/neurotechdevkit/rendering/_source.py
@@ -9,7 +9,7 @@ import scipy
 from matplotlib.image import BboxImage
 from matplotlib.transforms import Bbox, Transform, TransformedBbox
 
-from neurotechdevkit.sources import PhasedArraySource, Source
+from neurotechdevkit.sources import PhasedArraySource, PointSource, Source
 
 _COMPONENT_DIR = pathlib.Path(__file__).parent / "components"
 
@@ -43,6 +43,7 @@ class SourceDrawingParams(NamedTuple):
     aperture: float
     focal_length: float
     source_is_flat: bool
+    is_point_source: bool
 
 
 def create_source_drawing_artist(
@@ -61,6 +62,9 @@ def create_source_drawing_artist(
     Returns:
         A matplotlib artist containing the rendered source.
     """
+    if source_params.is_point_source:
+        return create_point_source_artist(source_params, transform)
+
     raw_img = _load_most_similar_source_image(
         source_params.aperture, source_params.focal_length, source_params.source_is_flat
     )
@@ -87,6 +91,35 @@ def create_source_drawing_artist(
     )
     img_box.set_data(transformed_img)
     return img_box
+
+
+def create_point_source_artist(
+    source_params: SourceDrawingParams, transform: Transform
+) -> matplotlib.artist.Artist:
+    """Create a matplotlib artist for a PointSource2D rendered inside a scenario.
+
+    Note that the source coordinates are in scenario coordinates, and not plot
+    coordinates.
+
+    Args:
+        source_params: the SourceDrawingParams that describe the source.
+        transform: A Transform function which maps from plot data coordinates into
+            display coordinates.
+
+    Returns:
+        A matplotlib artist containing the rendered source.
+    """
+    marker_size = 30
+    marker_style = "o"
+    artist = plt.scatter(
+        # plot x/y is flipped
+        source_params.position[1],
+        source_params.position[0],
+        s=marker_size,
+        marker=marker_style,
+        c="gray",
+    )
+    return artist
 
 
 def create_source_legend_artist(
@@ -176,6 +209,18 @@ def source_should_be_flat(source: Source) -> bool:
     """
     is_phased_array = isinstance(source, PhasedArraySource)
     return is_phased_array or np.isinf(source.focal_length)
+
+
+def source_should_be_point(source: Source) -> bool:
+    """Determine if a source should be represented as a point.
+
+    Args:
+        source: the source instance.
+
+    Returns:
+        True if the source is a PointSource, and False otherwise.
+    """
+    return isinstance(source, PointSource)
 
 
 def _select_image_file(

--- a/src/neurotechdevkit/results/_results.py
+++ b/src/neurotechdevkit/results/_results.py
@@ -281,6 +281,7 @@ class SteadyStateResult2D(SteadyStateResult):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 
@@ -382,6 +383,7 @@ class SteadyStateResult3D(SteadyStateResult):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 
@@ -780,6 +782,7 @@ class PulsedResult2D(PulsedResult):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 
@@ -1069,6 +1072,7 @@ class PulsedResult3D(PulsedResult):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 

--- a/src/neurotechdevkit/scenarios/_base.py
+++ b/src/neurotechdevkit/scenarios/_base.py
@@ -921,6 +921,7 @@ class Scenario2D(Scenario):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 
@@ -1101,6 +1102,7 @@ class Scenario3D(Scenario):
                     aperture=source.aperture,
                     focal_length=source.focal_length,
                     source_is_flat=rendering.source_should_be_flat(source),
+                    is_point_source=rendering.source_should_be_point(source),
                 )
                 rendering.draw_source(ax, drawing_params)
 

--- a/src/neurotechdevkit/sources.py
+++ b/src/neurotechdevkit/sources.py
@@ -153,6 +153,90 @@ class Source(abc.ABC):
         pass
 
 
+class PointSource(Source):
+    """Theoretical point source.
+
+    Automatically sets `aperture`, `focal_length`, and `direction`, and `num_points`.
+    """
+
+    def __init__(
+        self,
+        *,
+        position: npt.NDArray[np.float_],
+        num_dim: int = 2,
+        delay: float = 0.0,
+    ) -> None:
+        """Initialize a new point source."""
+        super().__init__(
+            position=position,
+            direction=np.full(shape=num_dim, fill_value=np.nan),
+            aperture=0.0,
+            focal_length=0.0,
+            num_points=1,
+            delay=delay,
+        )
+
+    def _calculate_coordinates(self) -> npt.NDArray[np.float_]:
+        """Calculate the coordinates of the point source cloud for the 2D source.
+
+        Returns:
+            An array containing the coordinates (in meters) of the point source.
+        """
+        return np.array([self.position])
+
+    def calculate_waveform_scale(self, dx: float) -> float:
+        """Calculate the scale factor to apply to waveforms from point source.
+
+        The scale is equal to the ratio between the density of source points
+        and the density of grid points (2D: 1 / dx; 3D: 1 / dx**2).
+        Because the aperture is technically zero, we approximate the source
+        density as the smallest grid (2D: 1 / dx; 3D: 1 / dx**2).
+
+        Args:
+            dx: the separation between gridpoints (in meters). Assumed to be the same in
+                both directions.
+                Unused.
+
+        Returns:
+            The scale factor to apply to the waveform.
+        """
+        return 1
+
+
+class PointSource2D(PointSource):
+    """A point source in 2D."""
+
+    def __init__(
+        self,
+        *,
+        position: npt.NDArray[np.float_],
+        delay: float = 0.0,
+    ) -> None:
+        """Initialize a new 2-D point source."""
+        super().__init__(
+            position=position,
+            num_dim=2,
+            delay=delay,
+        )
+
+
+class PointSource3D(PointSource):
+    """A point source in 3D."""
+
+    def __init__(
+        self,
+        *,
+        position: npt.NDArray[np.float_],
+        delay: float = 0.0,
+    ) -> None:
+        """Initialize a new 2-D point source."""
+        super().__init__(
+            position=position,
+            num_dim=3,
+            delay=delay,
+        )
+
+
 class FocusedSource2D(Source):
     """A focused source in 2D.
 
@@ -393,8 +477,8 @@ class FocusedSource3D(Source):
         return axis, theta
 
 
-class UnfocusedSource(Source):
-    """A base class for unfocused sources.
+class UnfocusedSource(Source, abc.ABC):
+    """An abstract base class for unfocused sources.
 
     Automatically sets `focal_length` to `np.inf`
     """

--- a/tests/neurotechdevkit/test_sources.py
+++ b/tests/neurotechdevkit/test_sources.py
@@ -12,6 +12,8 @@ from neurotechdevkit.sources import (
     PhasedArraySource3D,
     PlanarSource2D,
     PlanarSource3D,
+    PointSource2D,
+    PointSource3D,
     _rotate_2d,
     _rotate_3d,
 )
@@ -1719,3 +1721,55 @@ class TestPhasedArraySource3D(Source3DTestMixin):
         grid_density = 1 / dx**2
         point_density = quotient * num_elements / (length * height)
         np.testing.assert_allclose(scale, grid_density / point_density)
+
+
+class TestPointSource2D:
+    def test_init_properties(self):
+        """Verify that .position and .delay match the value received in the
+        constructor.
+        """
+        position = np.array([-2.0, 3.0])
+        delay = 123
+        source = PointSource2D(position=position, delay=delay)
+        np.testing.assert_allclose(source.position, position)
+        assert source.delay == delay
+
+    def test_calculate_coordinates(self):
+        """Verify that the calculated coordinates match the specified position."""
+        position = np.array([-2.0, 3.0])
+        source = PointSource2D(position=position)
+        np.testing.assert_allclose(
+            source._calculate_coordinates(), np.array([position])
+        )
+
+    def test_calculate_waveform_scale(self):
+        """Verify that calculate_waveform_scale returns 1."""
+        dx = 0.01
+        source = PointSource2D(position=np.array([-2.0, 3.0]))
+        assert source.calculate_waveform_scale(dx=dx) == 1
+
+
+class TestPointSource3D:
+    def test_init_properties(self):
+        """Verify that .position and .delay match the value received in the
+        constructor.
+        """
+        position = np.array([1.0, -2.0, 3.0])
+        delay = 123
+        source = PointSource3D(position=position, delay=delay)
+        np.testing.assert_allclose(source.position, position)
+        assert source.delay == delay
+
+    def test_calculate_coordinates(self):
+        """Verify that the calculated coordinates match the specified position."""
+        position = np.array([1.0, -2.0, 3.0])
+        source = PointSource3D(position=position)
+        np.testing.assert_allclose(
+            source._calculate_coordinates(), np.array([position])
+        )
+
+    def test_calculate_waveform_scale(self):
+        """Verify that calculate_waveform_scale returns 1."""
+        dx = 0.01
+        source = PointSource3D(position=np.array([1.0, -2.0, 3.0]))
+        assert source.calculate_waveform_scale(dx=dx) == 1

